README.md
# Idler
Periodically idles Kubernetes deployments meeting certain criteria to avoid wasting money

## How does it work?

The script performs the following steps:

1. For each [eligible][1] deployment in the cluster, across all namespaces
2. Calculate the [average CPU usage][4] of the deployment's containers over the
   last minute
3. If the usage is greater than a certain [percentage threshold][3], skip to
   the next deployment
4. Otherwise, *idle* the deployment, by:
   1. Adding a label and an annotation to the deployment
   2. Reducing the number of replicas to 0 (which deletes pods)
   3. Disabling the ingress for the deployment, and adding an ingress rule to
      the [unidler][2] ingress for the deployment's hostname (which redirects
      traffic to the [unidler][2] webapp)

## Testing

Build the docker image to run the tests:
```sh
docker build -t idler .
```

## Deployment

Deployed to the kubernetes cluster as a
[cronjob](https://github.com/ministryofjustice/analytics-platform-helm-charts/tree/master/charts/idler) using Helm


[1]: https://github.com/ministryofjustice/analytics-platform-idler/blob/master/idler.py#L109
[2]: https://github.com/ministryofjustice/analytics-platform-unidler
[3]: https://github.com/ministryofjustice/analytics-platform-idler/blob/master/idler.py#L36
[4]: https://github.com/ministryofjustice/analytics-platform-idler/blob/master/idler.py#L140


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [v0.5.2] - 2019-02-18
### Fixed
Updating `Service` correctly.

`Service` redirection needs to set `targetPort`.


## [v0.5.1] - 2019-02-18
### Fixed
Updating `Service` correctly.

There was a mistake in the code which redirected a Kubernetes `Service` to
the Unidler.


## [v0.5.0] - 2019-01-24
### Improvement
Redirect `Service` instead of `Ingress`.

Idling was achieved by disabling the app `Ingress` and adding the app hostname
to the unidler `Ingress`. If multiple users unidled at the same time, there
could be a race condition when removing the hostnames from the unidler `Ingress`.

This change does away with editing ingresses altogether and edits only the app
service. By changing the `Service` to what is essentially a `CNAME` forwarding to
the unidler, requests to the app host are handled by the unidler.


## [v0.4.0] - 2018-10-10
### Improvement
Lookup `Ingress` by App label.

For jupyter-lab support.


## [v0.3.1] - 2018-10-09
### Fixed
Add support for μcores


## [v0.3.0] - 2018-10-03
### Fixed
Support for metrics returned in nanocore units


## [v0.2.4] - 2018-04-27
### Improvements
Added more logging and improved consistency of logging.

[See PR](https://github.com/ministryofjustice/analytics-platform-idler/pull/11)


## [v0.2.3] - 2018-04-25
### Improvements
- Better handling of unhealthy pods (lacking metrics data)
- Fixed logging (and added support for `LOG_LEVEL` environment variable).

See PR: [#10](https://github.com/ministryofjustice/analytics-platform-idler/pull/10)


## [v0.2.2] - 2018-04-25
### Fixed
- Added missing file (`metrics_api.py`) to `Dockerfile`.


## [v0.2.1] - 2018-04-24
### Fixed
- Fixes idler incorrectly idling RStudio instances using more CPU than the threshold (`RSTUDIO_ACTIVIY_CPU_THRESHOLD`).


test/test_idler.py
from datetime import datetime, timezone
import json
from unittest.mock import MagicMock, patch

import pytest
from hypothesis import given, settings
from hypothesis.strategies import integers, text, composite

import idler
from idler import (
    IDLED,
    IDLED_AT,
    SERVICE_TYPE_EXTERNAL_NAME,
    UNIDLER_SERVICE_HOST,
)


@composite
def cpu_usage(draw):
    n = draw(integers(min_value=1))
    u = draw(text(alphabet='mnu', max_size=1, min_size=1))
    return f'{n}{u}'


@given(cpu_usage())
@settings(max_examples=1500)
def test_parse_cpuusage(cpu: str):
    out = 0
    if cpu.endswith('n'):
        out = int(cpu.rstrip('n')) / 1000000
    elif cpu.endswith('m'):
        out = int(cpu.rstrip('m'))
    elif cpu.endswith('u'):
        out = int(cpu.rstrip('u')) / 1000
    assert idler.core_val_with_unit_to_int(cpu) == out


@pytest.yield_fixture
def current_time():
    dt = MagicMock()
    now = datetime(2018, 2, 7, 11, 44, 20, tzinfo=timezone.utc)
    dt.now.return_value = now
    with patch('idler.datetime', dt):
        yield now


def mock_container(cpu_limit):
    container = MagicMock()
    container.resources.limits = {'cpu': cpu_limit}

    return container


@pytest.fixture
def deployment():
    deployment = MagicMock()
    deployment.metadata.annotations = {}
    deployment.metadata.labels = {'app': 'rstudio'}
    deployment.metadata.namespace = 'user-alice'
    deployment.spec.replicas = 2
    deployment.spec.template.spec.containers = [
        mock_container(cpu_limit='100m'),
        mock_container(cpu_limit='1500m'),
    ]
    return deployment


@pytest.fixture
def pod():
    pod = MagicMock()
    pod.metadata.name = 'rstudio-whatever-1234-abcde'
    pod.metadata.labels = {'app': 'rstudio'}
    pod.metadata.namespace = 'user-alice'
    return pod


@pytest.yield_fixture
def client(deployment, pod):
    client = MagicMock()
    apps_api = client.AppsV1beta1Api.return_value
    apps_api.list_deployment_for_all_namespaces.return_value.items = [
        deployment,
    ]

    core_api = client.CoreV1Api.return_value
    core_api.list_pod_for_all_namespaces.return_value.items = [
        pod,
    ]

    with patch('idler.client', client):
        yield client


@pytest.yield_fixture
def env():
    env = {}
    with patch('idler.os') as mock_os:
        mock_os.environ = env
        yield env


@pytest.fixture
def unidler():
    unidler = MagicMock()
    unidler.spec.rules = [
        MagicMock(),
    ]
    unidler.spec.tls = [MagicMock()]
    unidler.spec.tls[0].hosts = []
    return unidler


def mock_podmetric(cpu_usage=None):
    if cpu_usage is None:
        cpu_usage = ['0']
    metric = MagicMock(name='PodMetrics')
    metric.containers = []
    for usage in cpu_usage:
        container = MagicMock(name='Container')
        container.usage = {'cpu': usage}
        metric.containers.append(container)
    return metric


@pytest.yield_fixture
def metrics(deployment):
    metric = mock_podmetric()
    cache = {
        (deployment.metadata.labels['app'], deployment.metadata.namespace):
            metric,
    }
    with patch('idler.metrics_lookup', cache):
        yield cache


@pytest.yield_fixture
def pods_lookup(pod):
    cache = {
        (pod.metadata.labels['app'], pod.metadata.namespace): pod,
    }
    with patch('idler.pods_lookup', cache):
        yield cache


def test_idle_deployments(client, deployment, env, metrics):
    core_api = client.CoreV1Api.return_value
    apps_api = client.AppsV1beta1Api.return_value

    idler.idle_deployments()

    apps_api.patch_namespaced_deployment.assert_called_with(
        deployment.metadata.name,
        deployment.metadata.namespace,
        deployment)

    core_api.read_namespaced_service.assert_called_with(
        deployment.metadata.name,
        deployment.metadata.namespace)

    core_api.patch_namespaced_service.assert_called_with(
        name=deployment.metadata.name,
        namespace=deployment.metadata.namespace,
        body=[
            {
                "op": "replace",
                "path": "/spec/type",
                "value": SERVICE_TYPE_EXTERNAL_NAME},
            {
                "op": "add",
                "path": "/spec/externalName",
                "value": UNIDLER_SERVICE_HOST},
            {
                "op": "replace",
                "path": "/spec/ports",
                "value": [
                    {
                        "name": "http",
                        "port": 80,
                        "protocol": "TCP",
                        "targetPort": 80
                    }
                ]},
            {
                "op": "remove",
                "path": "/spec/selector"},
            {
                "op": "remove",
                "path": "/spec/clusterIP"}])


def test_eligible_deployments(client, env):
    deployments = idler.eligible_deployments()
    api = client.AppsV1beta1Api.return_value
    api.list_deployment_for_all_namespaces.assert_called_with(
        label_selector=f'!{IDLED},app=rstudio')
    assert len(list(deployments)) > 0


@pytest.mark.parametrize('label_selector, expected', [
    ('foo', f'!{IDLED},foo'),
    ('foo=bar', f'!{IDLED},foo=bar'),
    ('!foo', f'!{IDLED},!foo'),
    ('', f'!{IDLED}'),
])
def test_label_selector(client, env, label_selector, expected):
    env['LABEL_SELECTOR'] = label_selector
    idler.eligible_deployments()
    api = client.AppsV1beta1Api.return_value
    api.list_deployment_for_all_namespaces.assert_called_with(
        label_selector=expected)


def test_should_idle(deployment, env, metrics):
    assert idler.should_idle(deployment)


def test_should_not_idle(deployment, env):
    with patch('idler.avg_cpu_percent') as cpu:
        cpu.return_value = 100
        assert not idler.should_idle(deployment)


@pytest.mark.parametrize('cpu_usage, expected', [
    (['0'], 0),
    (['0', '0'], 0),
    (['100m', '0'], 6.25),  # 100 / (1500+100)
    (['100m', '100m'], 12.5),  # 200 / (1500+100)
    (['100m', '100000000n'], 12.5),  # mixed units # 200 / (1500+100)
    (['100000000n', '0'], 6.25),
    (['100000000n', '100m'], 12.5),
    (['100000000n', '100000000n'], 12.5),
    (['100m', '100000000n'], 12.5),
    (['100000u', '0'], 6.25),
    (['100000u', '100m'], 12.5),
    (['100000u', '100000000n'], 12.5),
    (['100m', '100000u'], 12.5),
])
def test_avg_cpu_percent(
        client, deployment, pods_lookup, metrics, cpu_usage, expected):
    key = (deployment.metadata.labels['app'], deployment.metadata.namespace)
    metrics[key] = mock_podmetric(cpu_usage)
    assert idler.avg_cpu_percent(deployment) == expected


def test_mark_idled(deployment, current_time):
    expected_replicas = deployment.spec.replicas

    idler.mark_idled(deployment)

    assert IDLED in deployment.metadata.labels
    assert IDLED_AT in deployment.metadata.annotations
    timestamp, replicas = deployment.metadata.annotations[IDLED_AT].split(',')
    assert int(replicas) == expected_replicas
    assert timestamp == current_time.isoformat(timespec='seconds')


def test_zero_replicas(deployment):
    idler.zero_replicas(deployment)

    assert deployment.spec.replicas == 0


def test_write_changes(client, deployment):
    idler.write_changes(deployment)

    api = client.AppsV1beta1Api.return_value
    api.patch_namespaced_deployment.assert_called_with(
        deployment.metadata.name,
        deployment.metadata.namespace,
        deployment
    )


k8s_oidc.py
import base64
import datetime
import json
import tempfile

import google.auth
import google.auth.transport.requests
import oauthlib.oauth2
import urllib3
from requests_oauthlib import OAuth2Session
from six import PY3

import kubernetes
from kubernetes.client import ApiClient, Configuration
from kubernetes.config.dateutil import UTC, format_rfc3339
from kubernetes.config.kube_config import (
    ConfigNode,
    FileOrData,
    _is_expired,
)


class KubeConfigLoader(object):

    def __init__(self, config_dict, active_context=None,
                 get_google_credentials=None,
                 config_base_path="",
                 config_persister=None):
        self._config = ConfigNode('kube-config', config_dict)
        self._current_context = None
        self._user = None
        self._cluster = None
        self.set_active_context(active_context)
        self._config_base_path = config_base_path
        self._config_persister = config_persister

        def _refresh_credentials():
            credentials, project_id = google.auth.default(
                scopes=['https://www.googleapis.com/auth/cloud-platform']
            )
            request = google.auth.transport.requests.Request()
            credentials.refresh(request)
            return credentials

        if get_google_credentials:
            self._get_google_credentials = get_google_credentials
        else:
            self._get_google_credentials = _refresh_credentials

    def set_active_context(self, context_name=None):
        if context_name is None:
            context_name = self._config['current-context']
        self._current_context = self._config['contexts'].get_with_name(
            context_name)
        if (self._current_context['context'].safe_get('user') and
                self._config.safe_get('users')):
            user = self._config['users'].get_with_name(
                self._current_context['context']['user'], safe=True)
            if user:
                self._user = user['user']
            else:
                self._user = None
        else:
            self._user = None
        self._cluster = self._config['clusters'].get_with_name(
            self._current_context['context']['cluster'])['cluster']

    def _load_authentication(self):
        """Read authentication from kube-config user section if exists.

        This function goes through various authentication methods in user
        section of kube-config and stops if it finds a valid authentication
        method. The order of authentication methods is:

            1. GCP auth-provider
            2. token_data
            3. token field (point to a token file)
            4. username/password
            4. oidc auth-provider
            5. username/password
        """
        if not self._user:
            return
        if self._load_gcp_token():
            return
        if self._load_user_token():
            return
        if self._load_oid_token():
            return
        self._load_user_pass_token()

    def _load_gcp_token(self):
        if 'auth-provider' not in self._user:
            return
        provider = self._user['auth-provider']
        if 'name' not in provider:
            return
        if provider['name'] != 'gcp':
            return

        if (('config' not in provider) or
                ('access-token' not in provider['config']) or
                ('expiry' in provider['config'] and
                 _is_expired(provider['config']['expiry']))):
            # token is not available or expired, refresh it
            self._refresh_gcp_token()

        self.token = "Bearer %s" % provider['config']['access-token']
        return self.token

    def _refresh_gcp_token(self):
        if 'config' not in self._user['auth-provider']:
            self._user['auth-provider'].value['config'] = {}
        provider = self._user['auth-provider']['config']
        credentials = self._get_google_credentials()
        provider.value['access-token'] = credentials.token
        provider.value['expiry'] = format_rfc3339(credentials.expiry)
        if self._config_persister:
            self._config_persister(self._config.value)

    def _load_oid_token(self):
        if 'auth-provider' not in self._user:
            return
        provider = self._user['auth-provider']

        if 'name' not in provider or 'config' not in provider:
            return

        if provider['name'] != 'oidc':
            return

        parts = provider['config']['id-token'].split('.')

        if len(parts) != 3:  # Not a valid JWT
            return None

        missing_padding = len(parts[1]) % 4
        if missing_padding != 0:
            parts[1] += '=' * (4 - missing_padding)

        if PY3:
            jwt_attributes = json.loads(
                base64.b64decode(parts[1]).decode('utf-8')
            )
        else:
            jwt_attributes = json.loads(
                base64.b64decode(parts[1] + "==")
            )

        expire = jwt_attributes.get('exp')

        if ((expire is not None) and
            (_is_expired(datetime.datetime.fromtimestamp(expire,
                                                         tz=UTC)))):
            self._refresh_oidc(provider)

            if self._config_persister:
                self._config_persister(self._config.value)

        self.token = "Bearer %s" % provider['config']['id-token']

        return self.token

    def _refresh_oidc(self, provider):
        ca_cert = tempfile.NamedTemporaryFile(delete=True)

        if PY3:
            cert = base64.b64decode(
                provider['config']['idp-certificate-authority-data']
            ).decode('utf-8')
        else:
            cert = base64.b64decode(
                provider['config']['idp-certificate-authority-data'] + "=="
            )

        with open(ca_cert.name, 'w') as fh:
            fh.write(cert)

        config = Configuration()
        config.ssl_ca_cert = ca_cert.name

        client = ApiClient(configuration=config)

        response = client.request(
            method="GET",
            url="%s/.well-known/openid-configuration"
            % provider['config']['idp-issuer-url']
        )

        if response.status != 200:
            return

        response = json.loads(response.data)

        request = OAuth2Session(
            client_id=provider['config']['client-id'],
            token=provider['config']['refresh-token'],
            auto_refresh_kwargs={
                'client_id': provider['config']['client-id'],
                'client_secret': provider['config']['client-secret']
            },
            auto_refresh_url=response['token_endpoint']
        )

        try:
            refresh = request.refresh_token(
                token_url=response['token_endpoint'],
                refresh_token=provider['config']['refresh-token'],
                auth=(provider['config']['client-id'],
                      provider['config']['client-secret']),
                verify=ca_cert.name
            )
        except oauthlib.oauth2.rfc6749.errors.InvalidClientIdError:
            return

        provider['config'].value['id-token'] = refresh['id_token']
        provider['config'].value['refresh-token'] = refresh['refresh_token']

    def _load_user_token(self):
        token = FileOrData(
            self._user, 'tokenFile', 'token',
            file_base_path=self._config_base_path,
            base64_file_content=False).as_data()
        if token:
            self.token = "Bearer %s" % token
            return True

    def _load_user_pass_token(self):
        if 'username' in self._user and 'password' in self._user:
            self.token = urllib3.util.make_headers(
                basic_auth=(self._user['username'] + ':' +
                            self._user['password'])).get('authorization')
            return True

    def _load_cluster_info(self):
        if 'server' in self._cluster:
            self.host = self._cluster['server']
            if self.host.startswith("https"):
                self.ssl_ca_cert = FileOrData(
                    self._cluster, 'certificate-authority',
                    file_base_path=self._config_base_path).as_file()
                self.cert_file = FileOrData(
                    self._user, 'client-certificate',
                    file_base_path=self._config_base_path).as_file()
                self.key_file = FileOrData(
                    self._user, 'client-key',
                    file_base_path=self._config_base_path).as_file()
        if 'insecure-skip-tls-verify' in self._cluster:
            self.verify_ssl = not self._cluster['insecure-skip-tls-verify']

    def _set_config(self, client_configuration):
        if 'token' in self.__dict__:
            client_configuration.api_key['authorization'] = self.token
        # copy these keys directly from self to configuration object
        keys = ['host', 'ssl_ca_cert', 'cert_file', 'key_file', 'verify_ssl']
        for key in keys:
            if key in self.__dict__:
                setattr(client_configuration, key, getattr(self, key))

    def load_and_set(self, client_configuration):
        self._load_authentication()
        self._load_cluster_info()
        self._set_config(client_configuration)

    def list_contexts(self):
        return [context.value for context in self._config['contexts']]

    @property
    def current_context(self):
        return self._current_context.value

kubernetes.config.kube_config.KubeConfigLoader = KubeConfigLoader


metrics_api.py
from pprint import pformat

from six import iteritems

import kubernetes
from kubernetes.client.api_client import ApiClient


class MetricsV1beta1Api(object):

    def __init__(self, api_client=None):
        if api_client is None:
            api_client = ApiClient()
        self.api_client = api_client

    def list_pod_metrics_for_all_namespaces(self, **kwargs):
        kwargs['_return_http_data_only'] = True
        if kwargs.get('async'):
            return self.list_pod_metrics_for_all_namespaces_with_http_info(
                **kwargs)
        else:
            (data) = self.list_pod_metrics_for_all_namespaces_with_http_info(
                **kwargs)
            return data

    def list_pod_metrics_for_all_namespaces_with_http_info(self, **kwargs):
        collection_formats = {}

        path_params = {}

        query_params = []
        if '_continue' in kwargs:
            query_params.append(('continue', kwargs['_continue']))
        if 'label_selector' in kwargs:
            query_params.append(('labelSelector', kwargs['label_selector']))

        form_params = []
        local_var_files = {}
        body_params = None

        header_params = {
            'Accept': self.api_client.select_header_accept([
                'application/json',
                'application/yaml',
                'application/vnd.kubernetes.protobuf',
                'application/json;stream=watch',
                'application/vnd.kubernetes.protobuf;stream=watch']),
            'Content-Type': self.api_client.select_header_content_type([
                '*/*']),
        }
        auth_settings = ['BearerToken']

        return self.api_client.call_api(
            f'/apis/metrics.k8s.io/v1beta1/pods',
            'GET',
            path_params,
            query_params,
            header_params,
            body=body_params,
            post_params=form_params,
            files=local_var_files,
            response_type='MetricsV1beta1PodMetricsList',
            auth_settings=auth_settings,
            async_req=kwargs.get('async_req'),
            _return_http_data_only=kwargs.get('_return_http_data_only'),
            _preload_content=kwargs.get('_preload_content', True),
            _request_timeout=kwargs.get('_request_timeout'),
            collection_formats=collection_formats)


kubernetes.client.apis.MetricsV1beta1Api = MetricsV1beta1Api
kubernetes.client.MetricsV1beta1Api = MetricsV1beta1Api


class Model(object):

    def to_dict(self):
        result = {}
        for attr, _ in iteritems(self.swagger_types):
            value = getattr(self, attr)
            if isinstance(value, list):
                result[attr] = list(map(
                    lambda x: x.to_dict() if hasattr(x, 'to_dict') else x,
                    value
                ))
            elif hasattr(value, 'to_dict'):
                result[attr] = value.to_dict()
            elif isinstance(value, dict):
                result[attr] = dict(map(
                    lambda item: (item[0], item[1].to_dict())
                    if hasattr(item[1], 'to_dict') else item,
                    value.items()
                ))
            else:
                result[attr] = value

        return result

    def to_str(self):
        return pformat(self.to_dict())

    def __repr__(self):
        return self.to_str()

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return False

        return self.__dict__ == other.__dict__

    def __ne__(self, other):
        return not self == other


class MetricsV1beta1PodMetricsList(Model):
    swagger_types = {
        'api_version': 'str',
        'items': 'list[MetricsV1beta1PodMetrics]',
        'kind': 'str',
        'metadata': 'V1ListMeta'
    }

    attribute_map = {
        'api_version': 'apiVersion',
        'items': 'items',
        'kind': 'kind',
        'metadata': 'metadata'
    }

    def __init__(self, api_version=None, items=None, kind=None, metadata=None):
        self.api_version = api_version
        self.items = items
        self.kind = kind
        self.metadata = metadata
        self.discriminator = None


kubernetes.client.models.MetricsV1beta1PodMetricsList = \
    MetricsV1beta1PodMetricsList
kubernetes.client.MetricsV1beta1PodMetricsList = MetricsV1beta1PodMetricsList


class MetricsV1beta1PodMetrics(Model):
    swagger_types = {
        'containers': 'list[MetricsV1beta1ContainerMetrics]',
        'metadata': 'V1ObjectMeta',
        'timestamp': 'datetime',
        'window': 'str',
    }

    attribute_map = {
        'containers': 'containers',
        'metadata': 'metadata',
        'timestamp': 'timestamp',
        'window': 'window',
    }

    def __init__(self, containers=None, metadata=None, timestamp=None, window=None):
        self.containers = containers
        self.metadata = metadata
        self.timestamp = timestamp
        self.window = window
        self.discriminator = None

kubernetes.client.models.MetricsV1beta1PodMetrics = MetricsV1beta1PodMetrics
kubernetes.client.MetricsV1beta1PodMetrics = MetricsV1beta1PodMetrics


class MetricsV1beta1ContainerMetrics(Model):
    swagger_types = {
        'name': 'str',
        'usage': 'dict(str, str)',
    }

    attribute_map = {
        'name': 'name',
        'usage': 'usage',
    }

    def __init__(self, name=None, usage=None):
        self.name = name
        self.usage = usage
        self.discriminator = None

kubernetes.client.models.MetricsV1beta1ContainerMetrics = MetricsV1beta1ContainerMetrics
kubernetes.client.MetricsV1beta1ContainerMetrics = MetricsV1beta1ContainerMetrics


idler.py
"""
Checks all RStudio deployments and idles those matching a kubernetes label
selector.
The label selector can be overridden by setting the LABEL_SELECTOR environment
variable.
See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
for label selector syntax.
"""

from datetime import datetime, timezone
import json
import logging
import os

from kubernetes import client, config

# provides swagger definitions for metrics api
import metrics_api


LOG_LEVEL = os.environ.get('LOG_LEVEL', 'INFO')

log = logging.getLogger(__name__)
log.setLevel(LOG_LEVEL)
log_formatter = logging.Formatter(
    '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
log_handler = logging.StreamHandler()
log_handler.setFormatter(log_formatter)
log.addHandler(log_handler)


ACTIVE_INSTANCE_CPU_PERCENTAGE = 90
try:
    ACTIVE_INSTANCE_CPU_PERCENTAGE = int(os.environ.get(
        'RSTUDIO_ACTIVITY_CPU_THRESHOLD', 90))
    log.debug(
        f'RSTUDIO_ACTIVITY_CPU_THRESHOLD={ACTIVE_INSTANCE_CPU_PERCENTAGE}%')
except ValueError:
    log.warning(
        'Invalid value for RSTUDIO_ACTIVITY_CPU_THRESHOLD, using default')

IDLED = 'mojanalytics.xyz/idled'
IDLED_AT = 'mojanalytics.xyz/idled-at'
SERVICE_TYPE_EXTERNAL_NAME = "ExternalName"
UNIDLER_SERVICE_HOST = "unidler.default.svc.cluster.local"


metrics_lookup = {}
pods_lookup = {}


def idle_deployments():
    build_lookups()

    for deployment in eligible_deployments():
        if should_idle(deployment):
            idle(deployment)


def get_key(pod_or_deployment):
    return (
        pod_or_deployment.metadata.labels['app'],
        pod_or_deployment.metadata.namespace,
    )


def build_metrics_lookup():
    metrics = client.MetricsV1beta1Api().list_pod_metrics_for_all_namespaces(
        label_selector=f'!{IDLED}{label_selector()}')

    for pod_metrics in metrics.items:
        pod_name = pod_metrics.metadata.name
        namespace = pod_metrics.metadata.namespace
        pod = pods_lookup[(pod_name, namespace)]
        app_name = pod.metadata.labels['app']

        metrics_lookup[(app_name, namespace)] = pod_metrics


def build_pods_lookup():
    pods = client.CoreV1Api().list_pod_for_all_namespaces(
        label_selector=f'!{IDLED}{label_selector()}')
    for pod in pods.items:
        pods_lookup[(pod.metadata.name, pod.metadata.namespace)] = pod


def build_lookups():
    build_pods_lookup()
    build_metrics_lookup()


def eligible_deployments():
    return client.AppsV1beta1Api().list_deployment_for_all_namespaces(
        label_selector=f'!{IDLED}{label_selector()}').items


def label_selector():
    label_selector = os.environ.get('LABEL_SELECTOR', 'app=rstudio')
    log.debug(f'LABEL_SELECTOR="{label_selector}"')
    if label_selector:
        label_selector = ',' + label_selector
    return label_selector


def should_idle(deployment):
    usage = 0
    key = get_key(deployment)

    try:
        usage = avg_cpu_percent(deployment)
    except ValueError as ve:
        log.exception(f'{key}: Using unknown unit of CPU: {ve}', exc_info=True)

    log.debug(f'{key}: Using {usage}% of CPU')

    if usage > ACTIVE_INSTANCE_CPU_PERCENTAGE:
        log.info(f'{key}: Not idling as using {usage}% of CPU')
        return False

    return True


def core_val_with_unit_to_int(core_val_with_unit: str):
    # millicpus have the 'm' suffix
    if core_val_with_unit.endswith('m'):
        return int(core_val_with_unit.rstrip('m'), 10)
    elif core_val_with_unit.endswith('n'):
        # nanocpus have the 'n' suffix
        return int(core_val_with_unit.rstrip('n'), 10) / 1000000
    elif core_val_with_unit.endswith('u'):
        # microcpus (μ)
        return int(core_val_with_unit.rstrip('u'), 10) / 1000
    else:
        # if the result is 0 then there is no suffix
        return int(core_val_with_unit, 10)


def avg_cpu_percent(deployment):
    key = get_key(deployment)
    try:
        metrics = metrics_lookup[key]
    except KeyError as e:
        log.warning(f'{key}: Metrics not found, pod may be unhealthy.')
        return 0

    usage = 0
    for container in metrics.containers:
        usage += core_val_with_unit_to_int(container.usage['cpu'])

    total = 0
    for container in deployment.spec.template.spec.containers:
        total += core_val_with_unit_to_int(container.resources.limits['cpu'])

    return usage / total * 100.0


def idle(deployment):
    mark_idled(deployment)
    svc = Service(deployment.metadata.name, deployment.metadata.namespace)
    svc.redirect_to_unidler()
    zero_replicas(deployment)
    write_changes(deployment)

    log.debug(f'{get_key(deployment)}: Idled')


def mark_idled(deployment):
    timestamp = datetime.now(timezone.utc).isoformat(timespec='seconds')
    deployment.metadata.labels[IDLED] = 'true'
    deployment.metadata.annotations[IDLED_AT] = (
        f'{timestamp},{deployment.spec.replicas}')


class Service(object):

    def __init__(self, name, namespace):
        self.name = name
        self.namespace = namespace
        self.service = client.CoreV1Api().read_namespaced_service(
            name, namespace)

    def patch(self, *patch):
        client.CoreV1Api().patch_namespaced_service(
            name=self.name,
            namespace=self.namespace,
            body=list(patch))

    def redirect_to_unidler(self):
        self.patch(
            {
                "op": "replace",
                "path": "/spec/type",
                "value": SERVICE_TYPE_EXTERNAL_NAME},
            {
                "op": "add",
                "path": "/spec/externalName",
                "value": UNIDLER_SERVICE_HOST},
            {
                "op": "replace",
                "path": "/spec/ports",
                "value": [
                    {
                        "name": "http",
                        "port": 80,
                        "protocol": "TCP",
                        "targetPort": 80
                    }
                ]},
            {
                "op": "remove",
                "path": "/spec/selector"},
            {
                "op": "remove",
                "path": "/spec/clusterIP"})


def zero_replicas(deployment):
    deployment.spec.replicas = 0


def write_changes(deployment):
    client.AppsV1beta1Api().patch_namespaced_deployment(
        deployment.metadata.name,
        deployment.metadata.namespace,
        deployment)


def load_kube_config():
    try:
        config.load_incluster_config()
    except config.ConfigException:
        # monkeypatch config loader to handle OIDC
        import k8s_oidc
        config.load_kube_config()


if __name__ == '__main__':
    load_kube_config()
    idle_deployments()
